### ISSUE TITLE
Malformed Boogie when comparing datatype with parent trait

### ISSUE THREAD

**Issue by RustanLeino** [dafny-lang/dafny#4994]
### Dafny version

4.4.0

### Code to produce this issue

```dafny
trait Parent {}

datatype D extends Parent = D

method Test(d: D, p: Parent) {
  assert p == d; // this works
  assert d == p; // this generates malformed Boogie
}
```

### Command to run and resulting output

```code
% dafny-4.4.0 verify test.dfy --general-traits:datatype --type-system-refresh
test.dfy(7,14): Error: invalid type for argument 0 in application of $IsA#_module.D: Box (expected: DatatypeType)
test.dfy(7,11): Error: invalid type for argument 0 in application of _module.D#Equal: Box (expected: DatatypeType)
test.dfy(7,14): Error: invalid type for argument 1 in application of _module.D#Equal: Box (expected: DatatypeType)
3 type checking errors detected in /var/folders/nf/22w8mpbd6bddjb72zlt3s_7rnr8bj2/T/test__module.bpl
```

### What happened?

The expression `p == d` translates correctly to Boogie, but `d == p` does not. Both are intended to work.

### What type of operating system are you experiencing the problem on?

Mac

**Comment by atomb** [comment]
This could be related to https://github.com/dafny-lang/dafny/issues/4983

**Comment by MikaelMayer** [comment]
Duplicate of https://github.com/dafny-lang/dafny/issues/4983

**Comment by RustanLeino** [comment]
This issue does indeed have the same symptoms as #4983, but it turns out that the reasons were different. Issue #4983 was fixed in PR #4997. The present issue is fixed in PR #5058.
